### PR TITLE
Using size ranges for e2e storage drivers and tests

### LINKED ISSUE #72080: E2E storage: dynamically determine test volume size
**What would you like to be added**:

`test/e2e/storage/testsuites` creates volumes dynamically. Initially, the size of those volumes was hard-coded in the test, which prevented using the tests with storage backends that couldn't support that hard-coded size.

https://github.com/kubernetes/kubernetes/pull/70862 introduces a mechanism where the driver for the storage backend can set the size.

But this still has the problem that the size chosen by the driver might not work for some tests. A better solution would be to let the driver and each test specify a size range, then when setting up the volume for a specific test, the actual size gets chosen by looking at the intersection.

**Why is this needed**:

Some tests might fail when the chosen size is too small or too large? Mostly a theoretic issue at this point, it hasn't happened in practice yet.

/kind feature

### PR DESCRIPTION
**Pull request by srmocher**
**What type of PR is this?**
/kind feature
/sig storage

**What this PR does / why we need it**:
Removes hardcoded sizes for volumes in tests and unifies the way sizes (or rather ranges) are defined for both the storage test drivers and the tests themselves. Both components specify a range of size to test (or in the case of the driver define what size range it supports). The intersection of the two ranges (the test and the driver's) are taken and the minimum of the result is chosen as the resource claim size.

Tests specify the size range as a property of `TestPattern` (is this the right way?). I'm not fully familiar with the size constraints on each storager driver yet (and in the tests - I've set some static values now, so added TODOs to remove them).

In the tests, we use `getSizeRangesIntersection` method and pass in the range from the `TestPattern` instance and the `driver` instance and take the intersection. 

**Which issue(s) this PR fixes**:

Fixes #72080 

**Special notes for your reviewer**:
I'm unfamiliar with the parameters to be set for each driver and the tests themselves (compared to the static values existing currently). Happy to know any pointers on setting these parameters.

**Does this PR introduce a user-facing change?**:
`NONE`

```release-note
NONE
```

### CODE REVIEW

**Comment by msau42** [comment]
cc @mkimuram

File: test/e2e/storage/testsuites/base.go [hunk 1]
```diff
@@ -344,3 +346,27 @@ func convertTestConfig(in *TestConfig) framework.VolumeTestConfig {
...
+// to be used as the claim size for the test.
+func getSizeRangesIntersection(first framework.SizeRange, second framework.SizeRange) string {
+	if strings.Compare(first.Units, second.Units) != 0 {
+		// both ranges must have same units
+		// TODO: return an error or try to convert the units instead
+		return ""
+	}
+
+	if second.Min > first.Max || first.Min > second.Max {
+		// TODO: this means a NULL intersection, return a better error
```
Comment (mkimuram) [inline test/e2e/storage/testsuites/base.go]: I think that making test fail here by calling ```framework.Failf``` will be good in this case.

File: test/e2e/storage/drivers/csi.go [hunk 2]
```diff
@@ -61,6 +61,11 @@ func initHostPathCSIDriver(name string, config testsuites.TestConfig, manifests
 			Name:        name,
 			FeatureTag:  "",
 			MaxFileSize: testpatterns.FileSizeMedium,
+			SupportedSizeRange: framework.SizeRange{
+				Min:   float64(1), // TODO: choose appropriate values
```
Comment (mkimuram) [inline test/e2e/storage/drivers/csi.go]: As for the values for ```Min``` and ```Max```,  we can safely choose below defaults:
[Each testsuite]
- Min: "1Mi"         ... Requested minimum requirement from driver side.
- Max: "10Gi"       ... More than or equal to known driver's maximum (5Gi).

[Each Driver]
- Min: "1Gi"          ... Original default size of the tests (Before increasing it to 5Gi)
- Max: "10Gi"       ... The same to testsuites' default maximum.

In addition, we need to add below exception:
[CSI GCE drivers]
- Min: "5Gi"          

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go#L173
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go#L267

File: test/e2e/storage/testsuites/testdriver.go [hunk 3]
```diff
@@ -72,10 +72,9 @@ type DynamicPVTestDriver interface {
 	// It will return nil, if the TestDriver doesn't support it.
 	GetDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
 
-	// GetClaimSize returns the size of the volume that is to be provisioned ("5Gi", "1Mi").
-	// The size must be chosen so that the resulting volume is large enough for all
-	// enabled tests and within the range supported by the underlying storage.
-	GetClaimSize() string
+	// GetClaimSize returns a range of size for which a volume can be provisioned with this driver.
+	// The actual size to be provisioned is determined dynamically based on this range and the test's range.
+	GetClaimSize() framework.SizeRange
```
Comment (mkimuram) [inline test/e2e/storage/testsuites/testdriver.go]: Should we rename it, because it now returns SizeRange, not size?

**Review by mkimuram — CHANGES_REQUESTED** [review]
@srmocher

Thank you for working on this. Changes look almost good.
Let me add some minor comments.

File: test/e2e/storage/testsuites/base.go [hunk 4]
```diff
@@ -344,3 +346,27 @@ func convertTestConfig(in *TestConfig) framework.VolumeTestConfig {
 		NodeSelector:   in.ClientNodeSelector,
 	}
 }
+
+// getSizeRangesIntersection takes two instances of storage size ranges and determines the
+// intersection of the intervals (if it exists) and return the minimum of the intersection
+// to be used as the claim size for the test.
+func getSizeRangesIntersection(first framework.SizeRange, second framework.SizeRange) string {
+	if strings.Compare(first.Units, second.Units) != 0 {
```
[the remark this answers is not in the document]
Comment (srmocher) [reply test/e2e/storage/testsuites/base.go]: Thank you for pointing that out - I wasn't aware of that method's availability. I will update the PR to use `ParseQuantity` and `Cmp` instead.

File: test/e2e/storage/testsuites/testdriver.go (hunk 3 above)
Comment (srmocher) [reply test/e2e/storage/testsuites/testdriver.go]: That makes sense - I shall update that as well.

**Comment by srmocher** [comment]
@mkimuram Could you PTAL? I've updated the code per your comments. 
I'm working on getting the CLA signed (issues with Linux ID).

**Comment by mkimuram** [comment]
@srmocher

Thank you for your fix. I confirmed that my comments are reflected perfectly.
So, ```getSizeRangesIntersection```'s logic looks good to me.

While I reviewed this PR again, I started to think that it would be more useful 
if the ```SupportedSizeRange``` for testsuites are defined in testsuites themselves,
instead of in testpatterns. Because ```SupportedSizeRange``` will depend more on
testsuites than testpatterns, and testsuites share testpatterns, as a result,
testpatterns which have size range inside might not be reusable by testsuites 
that support different size range.

Could you consider moving it from testpattern to testsuite?

**Comment by srmocher** [comment]
@mkimuram Sure, will update per your suggestion.

**Comment by srmocher** [comment]
Apologies for the late turnaround, I will push in rest of the fixes shortly.

**Pull request closed without merging**